### ISSUE TITLE
Problem adding attribute options to configurable product via REST Api

### ISSUE THREAD

**Issue by dimaportenko** [magento/magento2#9798]
### Preconditions
Magento 2.1.6

I'm trying to do following with Magento rest API
1. Create Configurable Product 
2. Create Simple Product
3. Add attribute options to Configurable Product
4. Add Simple Product to Configurable

### Steps to reproduce

I'm using following request to achieve this
```
<?php
$userData = array("username" => "dev", "password" => "q1w2e3r4");
$ch = curl_init("http://onepos.local/index.php/rest/V1/integration/admin/token/");

curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($userData));
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch, CURLOPT_HTTPHEADER, array("Content-Type: application/json", "Content-Lenght: " . strlen(json_encode($userData))));

$token = curl_exec($ch);
var_dump($token);

$sku = urlencode("Test-api-config-1");
$ch = curl_init("http://onepos.local/index.php/rest/V1/configurable-products/{$sku}/options/all");

curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "GET");
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch,  CURLOPT_HTTPHEADER, array("Content-Type: application/json", "Authorization: Bearer " . json_decode($token)));

$result = curl_exec($ch);
var_dump($result);

/// ---- configurable product
var_dump("/// ---- configurable product");
$post = '{
  "product":{
    "attribute_set_id":4,
    "created_at":null,    
    "custom_attributes":[
      {"attribute_code":"description",
      "value":"Lang verhaal over het artikel"},
      {"attribute_code":"short_description",
      "value":"kort verhaal over het artikel"},
      {"attribute_code":"required_options",
      "value":"1"},
      {"attribute_code":"options_container","value":"container2"},
      {"attribute_code":"has_options","value":"1"},
      {"attribute_code":"url_key","value":"Test-api-config-1"},
      {"attribute_code":"category_ids","value":["2","3"]},
      {"attribute_code":"manufacturer","value":1078}
    ],
    "extension_attributes":null,
    "id":0,
    "media_gallery_entries":null,
    "name":"Hello Word1",
    "options":null,"price":25.95,
    "product_links":null,
    "sku":"Test-api-config-1",
    "status":1,
    "tier_prices":null,
    "type_id":"configurable",
    "updated_at":null,
    "visibility":4,
    "weight":0
  }
}';

$child_sku = urlencode("Teun");
$ch = curl_init("http://onepos.local/index.php/rest/V1/products");

curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
curl_setopt($ch, CURLOPT_POSTFIELDS, $post);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch,  CURLOPT_HTTPHEADER, array("Content-Type: application/json", "Authorization: Bearer " . json_decode($token)));

$result = curl_exec($ch);
var_dump($result);

/// ---- Simple product
var_dump("/// ---- Simple product");
$post = '{
  "product":{
    "attribute_set_id":4,
    "created_at":null,
    "custom_attributes":[
      {"attribute_code":"description","value":"Lang verhaal over het artikel"},
      {"attribute_code":"short_description","value":"kort verhaal over het artikel"},
      {"attribute_code":"maat","value":6},
      {"attribute_code":"color","value":1068},
      {"attribute_code":"manufacturer","value":1078},
      {"attribute_code":"category_ids","value":["2","3"]}
    ],
    "extension_attributes":null,
    "id":0,
    "media_gallery_entries":null,
    "name":"Hello Word1",
    "options":null,
    "price":25.95,
    "product_links":null,
    "sku":"test-api-1",
    "status":1,
    "tier_prices":null,
    "type_id":"simple",
    "updated_at":null,
    "visibility":1,
    "weight":0
  }
}';

$child_sku = urlencode("Teun");
$ch = curl_init("http://onepos.local/index.php/rest/V1/products");

curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
curl_setopt($ch, CURLOPT_POSTFIELDS, $post);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch,  CURLOPT_HTTPHEADER, array("Content-Type: application/json", "Authorization: Bearer " . json_decode($token)));

$result = curl_exec($ch);
var_dump($result);

/// --- add options
var_dump("/// ---- add options");

$post = '{
  "option": {
    "id": 0,
    "attributeId": "90",
    "label": "Color",
    "position": 0,
    "is_use_default": true,
    "values": [
      {
        "valueIndex": 1068,
        "extension_attributes": {}
      }
    ],
    "extensionAttributes": {},
    "productId": 1449
  },

  "option": {
    "id": 0,
    "attributeId": "131",
    "label": "Maat",
    "position": 1,
    "values": [
      {
        "valueIndex": 6
      }
    ],
    "productId": 1449
  }
}';

$sku = urlencode("Test-api-config-1");
$ch = curl_init("http://onepos.local/rest/V1/configurable-products/{$sku}/options");
curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
curl_setopt($ch, CURLOPT_POSTFIELDS, $post);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch,  CURLOPT_HTTPHEADER, array("Content-Type: application/json", "Authorization: Bearer " . json_decode($token)));

$result = curl_exec($ch);
var_dump($result);
/// --- add child product
var_dump("/// ---- add child product");
$post = '
{
    "childSku": "test-api-1"
}
';
$sku = urlencode("Test-api-config-1");
$ch = curl_init("http://onepos.local/index.php/rest/V1/configurable-products/{$sku}/child");
curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
curl_setopt($ch, CURLOPT_POSTFIELDS, $post);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch,  CURLOPT_HTTPHEADER, array("Content-Type: application/json", "Authorization: Bearer " . json_decode($token)));

$result = curl_exec($ch);
var_dump($result);

```

### Expected result

1. Options are added to a configurable product and me able to add simple as a child to configurable.

### Actual result

1. On add options step I've recieve following error

> string(1369) "{"message":"Something went wrong while saving option.","trace":"#0 [internal function]: Magento\\ConfigurableProduct\\Model\\OptionRepository->save('Test-api-config...', Object(Magento\\ConfigurableProduct\\Model\\Product\\Type\\Configurable\\Attribute))\n#1 \/var\/www\/4onepos.mage2redis.interactivated.me\/httpdocs\/vendor\/magento\/module-webapi\/Controller\/Rest.php(307): call_user_func_array(Array, Array)\n#2 \/var\/www\/4onepos.mage2redis.interactivated.me\/httpdocs\/vendor\/magento\/module-webapi\/Controller\/Rest.php(216): Magento\\Webapi\\Controller\\Rest->processApiRequest()\n#3 \/var\/www\/4onepos.mage2redis.interactivated.me\/httpdocs\/var\/generation\/Magento\/Webapi\/Controller\/Rest\/Interceptor.php(37): Magento\\Webapi\\Controller\\Rest->dispatch(Object(Magento\\Framework\\App\\Request\\Http))\n#4 \/var\/www\/4onepos.mage2redis.interactivated.me\/httpdocs\/vendor\/magento\/framework\/App\/Http.php(135): Magento\\Webapi\\Controller\\Rest\\Interceptor->dispatch(Object(Magento\\Framework\\App\\Request\\Http))\n#5 \/var\/www\/4onepos.mage2redis.interactivated.me\/httpdocs\/vendor\/magento\/framework\/App\/Bootstrap.php(258): Magento\\Framework\\App\\Http->launch()\n#6 \/var\/www\/4onepos.mage2redis.interactivated.me\/httpdocs\/pub\/index.php(37): Magento\\Framework\\App\\Bootstrap->run(Object(Magento\\Framework\\App\\Http))\n#7 {main}"}"

I've debugged it and found following error 
`SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '455' for key 'PRIMARY', query was: INSERT INTO `catalog_product_super_attribute_label` (`product_super_attribute_id`,`use_default`,`store_id`,`value`) VALUES (?, ?, ?, ?) ON DUPLICATE KEY UPDATE `value` = VALUES(`value`), `use_default` = VALUES(`use_default`)`

vendor/magento/framework/Model/ResourceModel/Db/AbstractDb.php line 408. 
Magento 2.1.6

**Comment by teunlielu** [comment]
Work arrond: Remove: "id":0,
{"option":{"attribute_id":"131","extensionAttributes":null,"isUseDefault":false,"label":"maat","position":0,"product_id":1475,"values":[{"extensionAttributes":null,"value_index":6}]}}

**Comment by magento-engcom-team** [comment]
@troublediehard, thank you for your report.
We've created internal ticket(s) MAGETWO-80638 to track progress on the issue.

**Comment by DanyelleJean** [comment]
Has this ever been fixed?

**Comment by slacka1** [comment]
In 2.2.3 when you save OptionInterface it throws no errors and seems to clear the values() member. I believe it's save routine comes from: vendor/magento/framework/Model/AbstractModel.php.  The comments above the save routine say it's depreciated. Maybe this is also causing issues with the REST api, assuming it saves using this routine. Below is my code:
https://community.magento.com/t5/Magento-2-x-Programming/How-to-add-ConfigurableProducts-in-Magento-2-2-3/td-p/113681

**Comment by lpouwelse** [comment]
OptionId in not loaded / set for already existing configurable options, so a duplicate entry is created. Patch attached based on 2.3.1.
[9798.txt](https://github.com/magento/magento2/files/3339060/9798.txt)

.

**Comment by magento-engcom-team** [comment]
Hi @troublediehard. Thank you for your report. 
The issue has been fixed in magento/magento2#23529 by @lpouwelse in 2.3-develop branch
Related commit(s):
 - [f6c957cb0c92776f4dc10a0ebc1c9a1b4884db51](https://github.com/magento/magento2/commit/f6c957cb0c92776f4dc10a0ebc1c9a1b4884db51)
 - [c0a360fadc542dabead1425244d5080321328f1c](https://github.com/magento/magento2/commit/c0a360fadc542dabead1425244d5080321328f1c)

The fix will be available with the upcoming 2.3.3 release.
